docs/guides/advanced/authentication.md

```diff
@@ -305,7 +305,38 @@ However if the API provider deviates too far from the OAuth 2.0 specification it
     Coda doesn't currently support the older 1.0 or 1.0a versions of the OAuth specification. If you would like to connect to an API that only supports these versions of the standard please [contact support][support] so that we can continue to gauge interest.
 
 
-<!-- TODO: Progressive OAuth -->
+#### Incremental OAuth
+
+If your Pack includes a lot of features you may likewise need to request a lot of scopes during authentication, which can scare away new users. An alternative approach is to use incremental authorization, where you start out only requesting a small initial set of scopes and request more as the user starts using features that require them.
+
+Incremental authorization is made possible in Packs using the formula field [`extraOAuthScopes`][extraOAuthScopes]. You can use it to specify additional scopes that are needed in order to run that specific formula.
+
+```ts
+pack.setUserAuthentication({
+  type: coda.AuthenticationType.OAuth2,
+  // ...
+  scopes: ["read"],
+});
+
+// ...
+
+pack.addFormula({
+  name: "UpdateItem",
+  // ...
+  isAction: true,
+  extraOAuthScopes: ["update"],
+  // ...
+});
+```
+
+When the Pack above is installed the user will only be required to grant access to the `read` scope. However, when they try to use the `UpdateItem` action formula they will then be prompted grant additional access to the `write` scope. This prompt is displayed as a pop-up dialog at the bottom of the doc:
+
+<img src="../../../images/auth_oauth_incremental.png" srcset="../../../images/auth_oauth_incremental_2x.png 2x" class="screenshot" alt="Prompting the user for additional permissions">
+
+When the user signs in again they will be prompted to approve the additional scopes, after which they will be able to use the formula successfully.
+
+!!! info "401 response required"
+    Even if a formula specifies `extraOAuthScopes`, Coda will first try to run the formula using the existing credentials. The user will only be prompted to approve additional scopes if the formula fails with a 401 "Unauthorized" response from the API. If the API you are connecting to returns a different response code on failure you won't be able to take advantage of the incremental authorization feature.
```

Comment:
FWIW the Figma API incorrectly returns 403 and so our pack hacks the responses so they are treated as 401s, otherwise OAuth refreshes would not work for Figma (bonkers that this is how their API works). So it is possible, but not sure if we want to talk about that.

https://github.com/coda/packs/blob/main/packs/figma/main/formulas.ts#L97

I guess if it comes up enough we can expose an SDK helper function that abstracts this away.

Reply:
Technically, today we don't pay attention to the status code at all - any and all errors from a formula execution are clobbered if we notice there are missing scopes. We have this bug to change that: https://golinks.io/bug/19751 - maybe we should bump it up to a P2?

Reply:
Oh, good to know. Until we fix that bug I've just removed this warning block.